### DISCUSSION TITLE
Indicate grid internal loading/calculating state

### DISCUSSION THREAD

**Discussion by githorse** [KevinVandy/material-react-table#522] [Ideas]
The [loading stuff](https://www.material-react-table.com/docs/guides/async-loading#async-loading-feature-guide) is super cool, but it does not capture aggregations or other expensive grid calculations that cause the app to be unresponsive and the data to be (briefly) stale. For instance, I'm playing locally with a grid with 50k rows and a grouped-by `sum` aggregation. At 50k rows, the aggregation time after I reload new data into the grid) is around one second, during which time the grid's numbers are out of date, the screen is frozen, and there's no way (that I can see) to indicate to the user that something is still "loading" (calculating). Same issue if I have, say, 100k rows and I hit the "Sort" button.

I have exactly [the same issue with the MUI X grid](https://github.com/mui/mui-x/issues/8197). My suggestions here are similar:

1. provide some kind of callback like [`onContentReady`](https://js.devexpress.com/Documentation/18_1/ApiReference/UI_Widgets/dxDeferRendering/Configuration/#onContentReady) that fires when the grid is fully calculated, all data is current, and the screen has redrawn. This allows me to, say, display a loading spinner (or dim the opacity or whatever) while the grid is calculating.
2. include some sort of flag somewhere (`showCalculatingAsLoading`) that indicates that the grid should display as loading while calculation is happening. (This is tricky because I probably don't want to show skeletons during this time, but rather just dim the opacity or something more subtle. It would be great to distinguish "loading" (no data) from "reloading/calculating" (stale data) for this reason.)

**Comment by KevinVandy** [comment]
you can try to use `showProgressBars`, but since these calculations block the main thread, there will be pauses. Not quite sure what your suggesting with the callback.

**Comment by githorse** [comment]
Previously I've used the [DevExtreme datagrid](https://js.devexpress.com/), which has an [`onContentReady` callback](https://js.devexpress.com/Documentation/18_1/ApiReference/UI_Widgets/dxDeferRendering/Configuration/#onContentReady) that's called whenever data is reloaded (or re-aggregated or whatever) and the grid is ready to re-render. This is very useful because I can fully control the "loading" state of my page: whether it's loading data from an API or calculating something in the grid, I want to indicate to my user that the data is stale / the page is briefly unavailable / something is happening right now. So, for instance, I've used code like roughly this before:

```
const [loading, setLoading] = useState(false)

reload() {
  setLoading(true)
  fetch('some-url').then(setData)
}

return (
  <div>
    <button>Reload</Button>
    <Grid
      data={data}
      ...{someAggregationHere}
      onContentReady={() => setLoading(false)}
    />
  <div/>
)
```

The progress bars are very cool, but in general I want the freedom to express the loading state however I want. (Two solutions I've used in the past are (1) blocking the whole page with a modal loading overlay and (2) just dimming the opacity on the grid.)
